File: frontend/shader/contract/ts/dogecoin.ts
import { JsonRpcResponse } from "../app/app";

// ============================================================================
// Dogecoin Configuration
// ============================================================================

export interface DogecoinConfig {
  rpcUrl: string;
  rpcUser: string;
  rpcPassword: string;
  network: "mainnet" | "testnet";
}

/**
 * Dogecoin Connector
 */
export class DogecoinConnector {
  private readonly config: DogecoinConfig;

  constructor(config: DogecoinConfig) {
    this.config = config;
  }

  async getBalance(address: string): Promise<string> {
    if (!this.isValidAddress(address)) {
      throw new Error("Invalid Dogecoin address");
    }

    try {
      // Try RPC first if configured
      if (
        this.config.rpcUrl &&
        this.config.rpcUser &&
        this.config.rpcPassword
      ) {
        const response = await fetch(this.config.rpcUrl, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Basic ${Buffer.from(
              `${this.config.rpcUser}:${this.config.rpcPassword}`
            ).toString("base64")}`,
          },
          body: JSON.stringify({
            jsonrpc: "2.0",
            id: 1,
            method: "getreceivedbyaddress",
            params: [address, 0],
          }),
        });

        if (response.ok) {
          const data = (await response.json()) as JsonRpcResponse;
          if (data.result !== undefined) {
            // Convert from DOGE to doges (smallest unit, 1 DOGE = 100,000,000 doges)
            return (
              Number.parseFloat(data.result as string) * 100000000
            ).toString();
          }
        }
      }

      // Fallback to public API
      const apiUrl = `https://dogechain.info/api/v1/address/balance/${address}`;
      const response = await fetch(apiUrl);

      if (response.ok) {
        const data = (await response.json()) as { balance?: string };
        // Return balance in doges
        return (Number.parseFloat(data.balance || "0") * 100000000).toString();
      }

      throw new Error("Failed to fetch Dogecoin balance");
    } catch (error) {
      throw new Error(
        `Dogecoin getBalance failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  async sendTransaction(
    fromAddress: string,
    toAddress: string,
    amount: string
  ): Promise<string> {
    if (!this.isValidAddress(fromAddress) || !this.isValidAddress(toAddress)) {
      throw new Error("Invalid Dogecoin address");
    }

    const amountNum = Number.parseInt(amount, 10);
    if (Number.isNaN(amountNum) || amountNum <= 0) {
      throw new Error("Invalid amount");
    }

    try {
      // Similar to Bitcoin/Litecoin - requires RPC with wallet
      if (
        !this.config.rpcUrl ||
        !this.config.rpcUser ||
        !this.config.rpcPassword
      ) {
        throw new Error(
          "Dogecoin RPC credentials required for sending transactions"
        );
      }

      // Get UTXOs
      const utxoResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Basic ${Buffer.from(
            `${this.config.rpcUser}:${this.config.rpcPassword}`
          ).toString("base64")}`,
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 1,
          method: "listunspent",
          params: [0, 9999999, [fromAddress]],
        }),
      });

      if (!utxoResponse.ok) {
        throw new Error("Failed to fetch UTXOs");
      }

      const utxoData = (await utxoResponse.json()) as JsonRpcResponse;
      if (!utxoData.result || (utxoData.result as any[]).length === 0) {
        throw new Error("No UTXOs available");
      }

      // Create, sign, and send transaction
      const createTxResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Basic ${Buffer.from(
            `${this.config.rpcUser}:${this.config.rpcPassword}`
          ).toString("base64")}`,
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 2,
          method: "createrawtransaction",
          params: [
            (utxoData.result as any[]).map((utxo: any) => ({
              txid: utxo.txid,
              vout: utxo.vout,
            })),
            { [toAddress]: amountNum / 100000000 },
          ],
        }),
      });

      if (!createTxResponse.ok) {
        throw new Error("Failed to create transaction");
      }

      const createTxData = (await createTxResponse.json()) as JsonRpcResponse;
      const rawTx = createTxData.result as string;

      // Sign and send
      const signTxResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Basic ${Buffer.from(
            `${this.config.rpcUser}:${this.config.rpcPassword}`
          ).toString("base64")}`,
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 3,
          method: "signrawtransactionwithwallet",
          params: [rawTx],
        }),
      });

      if (!signTxResponse.ok) {
        throw new Error("Failed to sign transaction");
      }

      const signTxData = (await signTxResponse.json()) as JsonRpcResponse;
      const signedTx = (signTxData.result as { hex: string }).hex;

      const sendTxResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Basic ${Buffer.from(
            `${this.config.rpcUser}:${this.config.rpcPassword}`
          ).toString("base64")}`,
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 4,
          method: "sendrawtransaction",
          params: [signedTx],
        }),
      });

      if (!sendTxResponse.ok) {
        throw new Error("Failed to send transaction");
      }

      const sendTxData = (await sendTxResponse.json()) as JsonRpcResponse;
      return sendTxData.result as string;
    } catch (error) {
      throw new Error(
        `Dogecoin sendTransaction failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  isValidAddress(address: string): boolean {
    return this.config.network === "mainnet"
      ? address.startsWith("D")
      : address.startsWith("n") || address.startsWith("m");
  }

  /**
   * Bridge tokens from Dogecoin to another chain
   */
  async bridgeTokens(
    sourceAddress: string,
    destinationAddress: string,
    amount: string
  ): Promise<{ sourceTx: string; destinationTx?: string }> {
    const txHash = await this.sendTransaction(
      sourceAddress,
      destinationAddress,
      amount
    );
    return { sourceTx: txHash };
  }

  /**
   * Get chain name for bridge operations
   */
  getChainName(): string {
    return "Dogecoin";
  }

  /**
   * Get chain aliases for bridge operations
   */
  getChainAliases(): string[] {
    return ["dogecoin", "doge"];
  }
}

File: store/rust/rust/dogecoin.rs
//! Dogecoin connector (similar to Bitcoin)

use crate::{bitcoin_rpc_call, BridgeError, BridgeResult, DogecoinNetwork};
use serde::{Deserialize, Serialize};

/// Dogecoin connector (similar to Bitcoin)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DogecoinConnector {
    rpc_url: String,
    rpc_user: String,
    rpc_password: String,
    #[allow(dead_code)]
    network: DogecoinNetwork,
}

impl DogecoinConnector {
    pub fn new(rpc_url: &str, rpc_user: &str, rpc_password: &str, network: DogecoinNetwork) -> BridgeResult<Self> {
        Ok(Self {
            rpc_url: rpc_url.to_string(),
            rpc_user: rpc_user.to_string(),
            rpc_password: rpc_password.to_string(),
            network,
        })
    }

    pub fn get_balance(&self, address: &str) -> BridgeResult<u64> {
        if !self.is_valid_address(address) {
            return Err(BridgeError::InvalidAddress(address.to_string()));
        }

        // Dogecoin uses same RPC as Bitcoin
        let params = serde_json::json!([address]);
        let result = bitcoin_rpc_call(&self.rpc_url, &self.rpc_user, &self.rpc_password, "listunspent", params)?;

        let utxos: Vec<serde_json::Value> = serde_json::from_value(result)
            .map_err(|e| BridgeError::DeserializationError(format!("Failed to parse UTXOs: {}", e)))?;

        let balance: u64 = utxos
            .iter()
            .filter_map(|utxo| {
                utxo.get("amount")
                    .and_then(|a| a.as_f64())
                    .map(|amt| (amt * 100_000_000.0) as u64)
            })
            .sum();

        Ok(balance)
    }

    pub fn send_transaction(
        &self,
        from_address: &str,
        to_address: &str,
        amount: u64, // in smallest unit
    ) -> BridgeResult<String> {
        if !self.is_valid_address(from_address) || !self.is_valid_address(to_address) {
            return Err(BridgeError::InvalidAddress("Invalid Dogecoin address".to_string()));
        }
        if amount == 0 {
            return Err(BridgeError::InvalidAmount("Amount must be greater than 0".to_string()));
        }

        // Dogecoin uses same RPC as Bitcoin
        let amount_doge = amount as f64 / 100_000_000.0;
        let params = serde_json::json!([to_address, amount_doge]);
        let result = bitcoin_rpc_call(&self.rpc_url, &self.rpc_user, &self.rpc_password, "sendtoaddress", params)?;

        let tx_hash = result
            .as_str()
            .ok_or_else(|| BridgeError::DeserializationError("Invalid transaction hash format".to_string()))?
            .to_string();

        Ok(tx_hash)
    }

    pub fn is_valid_address(&self, address: &str) -> bool {
        // Dogecoin addresses start with 'D' for mainnet
        match self.network {
            DogecoinNetwork::Mainnet => address.starts_with('D'),
            DogecoinNetwork::Testnet => address.starts_with('n') || address.starts_with('m'),
        }
    }
}
